includes/class-amp-service-worker.php

```diff
@@ -0,0 +1,316 @@
+		}
+		if ( true === $theme_support['service_worker'] || ! empty( $theme_support['service_worker']['google_fonts_caching'] ) ) {
+			add_action( 'wp_front_service_worker', array( __CLASS__, 'add_google_fonts_caching' ) );
+		}
+	}
+
+	/**
+	 * Add query var for iframe service worker request.
+	 *
+	 * @param array $vars Query vars.
+	 * @return array Amended query vars.
+	 */
+	public static function add_query_var( $vars ) {
+		$vars[] = self::INSTALL_SERVICE_WORKER_IFRAME_QUERY_VAR;
+		return $vars;
+	}
+
+	/**
+	 * Add runtime caching for scripts loaded from the AMP CDN with a stale-while-revalidate strategy.
+	 *
+	 * @link https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js
+	 *
+	 * @param WP_Service_Worker_Scripts $service_workers Service worker registry.
+	 */
+	public static function add_cdn_script_caching( $service_workers ) {
+		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
+			_doing_it_wrong( __METHOD__, esc_html__( 'Expected argument to be WP_Service_Worker_Cache_Registry.', 'amp' ), '1.1' );
+			return;
+		}
+
+		// Add AMP scripts to runtime cache which will then get stale-while-revalidate strategy.
+		$service_workers->register(
+			'amp-cdn-runtime-caching',
+			function() {
+				$urls = AMP_Service_Worker::get_precached_script_cdn_urls();
+				if ( empty( $urls ) ) {
+					return '';
+				}
+
+				$js = file_get_contents( AMP__DIR__ . '/assets/js/amp-service-worker-runtime-precaching.js' ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents, WordPress.WP.AlternativeFunctions.file_system_read_file_get_contents
+				$js = preg_replace( '#/\*\s*global.+?\*/#', '', $js );
+				$js = str_replace(
+					'URLS',
+					wp_json_encode( $urls ),
+					$js
+				);
+				return $js;
+			}
+		);
+
+		// Serve the AMP Runtime from cache and check for an updated version in the background. See <https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js#L54-L58>.
+		$service_workers->caching_routes()->register(
+			'^https:\/\/cdn\.ampproject\.org\/.*',
+			array(
+				'strategy' => WP_Service_Worker_Caching_Routes::STRATEGY_STALE_WHILE_REVALIDATE,
+			)
+		);
+	}
+
+	/**
+	 * Add runtime image caching from the origin with a cache-first strategy.
+	 *
+	 * @link https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js#L60-L74
+	 *
+	 * @param WP_Service_Worker_Scripts $service_workers Service workers.
+	 */
+	public static function add_image_caching( $service_workers ) {
+		if ( ! ( $service_workers instanceof WP_Service_Worker_Scripts ) ) {
+			_doing_it_wrong( __METHOD__, esc_html__( 'Expected argument to be WP_Service_Worker_Scripts.', 'amp' ), '1.1' );
```

Comment:
See above. How about replacing with a type hint in the next release?

Doing it this way also gives us more flexibility to modify the PWA plugin's API without causing a fatal error in the AMP plugin.

Reply:
Same as above. I'd argue we should remove the `_doing_it_wrong()` because nobody is really doing anything wrong, they're just on an old version.

Because of that, I'd say we wait a little further with a type hint, just to decrease the chances of somebody using PWA plugin 0.1.